File: src/main/GO/leetcode/lists/topinterview/LeetCode350.go
package topinterview

func intersect(nums1 []int, nums2 []int) []int {
	arr1, arr2 := make([]int, 1001), make([]int, 1001)
	for i := 0; i < len(nums1); i++ {
		arr1[nums1[i]]++
	}
	for i := 0; i < len(nums2); i++ {
		arr2[nums2[i]]++
	}
	list := make([]int, 0)
	for i := 0; i < 1001; i++ {
		if arr1[i] > 0 && arr2[i] > 0 {
			for j := 0; j < Min(arr1[i], arr2[i]); j++ {
				list = append(list, i)
			}
		}
	}
	return list
}

File: src/main/java/leetcode/lists/topinterview/LeetCode350.java
package leetcode.lists.topinterview;

import java.util.ArrayList;
import java.util.List;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/29 17:14
 */
public class LeetCode350 {

    public int[] intersect(int[] nums1, int[] nums2) {
        int[] freq1 = getFreq(nums1);
        int[] freq2 = getFreq(nums2);

        List<Integer> ans = new ArrayList<>();

        for (int i = 0; i < 1001; i++) {
            int count = Math.min(freq1[i], freq2[i]);
            while (count-- > 0) {
                ans.add(i);
            }
        }

        return ans.stream().mapToInt(i -> i).toArray();
    }

    private int[] getFreq(int[] arr) {
        int[] freq = new int[1001];
        for (int i : arr) freq[i]++;
        return freq;
    }
}
